### PR TITLE
Normalize transport internal events

### PR DESCRIPTION
**Pull request by fcanela**
Previous code triggered a custom event per transport. This simplifies the code by providing a single logging event, plus providing details about the missing environment v
ariables.

## Summary by CodeRabbit

- **New Features**
  - Added Telegram and Email integration for notifications alongside existing Slack support.
  - Introduced a new SMTP service in Docker setup for email notifications.

- **Enhancements**
  - Updated environment configuration to include Telegram and SMTP settings.
  - Improved bot event handling to support multiple transport mechanisms (Slack, Telegram, Email).

- **Tests**
  - Expanded test coverage for Telegram and Email services.
  - Added mock data and utilities to support new notification channels in tests.

- **Documentation**
  - Updated example environment file with new variables for Telegram and SMTP configuration.

### CODE REVIEW

**Comment by jaybuidl** [comment]
@coderabbitai full review

File: src/utils/env.ts [hunk 1]
```diff
@@ -67,9 +67,57 @@ export const schema = {
...
+  SMTP_FROM: email({
+    desc: "Sender email address",
+    example: "no-reply@kleros.local",
+    default: undefined,
+  }),
+  SMTP_TO: email({
+    desc: "Sender email address",
+    example: "alert@kleros.local",
+    default: undefined,
+  }),
```
[the remark this answers is not in the document]
Comment (fcanela) [reply src/utils/env.ts:116]: Those env vars are expected to have a value or be `undefined`. They are defining transports that are optional.

Transports have checks like:
```ts
  const requiredFields: Array<keyof Env> = [
    "SMTP_HOST",
    "SMTP_PORT",
    "SMTP_SECURE",
    "SMTP_USER",
    "SMTP_PASSWORD",
    "SMTP_FROM",
    "SMTP_TO",
  ];
  const missingFields = requiredFields.filter((varName) => env[varName] === undefined);
  if (missingFields.length > 0) {
    emitter.emit(BotEventNames.TRANSPORT_CONFIGURATION_MISSING, {
      name: TRANSPORT_NAME,
      missing: missingFields,
    });
    return;
  }
```

Those checks ensures the required env vars are defined. If they are not, issues a warning with the missing deps.

File: src/services/email.ts [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+  message: EmailMessage;
+  from: Env["SMTP_FROM"];
+  to: Env["SMTP_TO"];
+  sendFn?: Transporter["sendMail"];
+};
+
+export const configurableSend = async (deps: ConfigurableSendDeps) => {
+  const { from, to, message, sendFn } = deps;
+
+  if (!sendFn) return;
```
[the remark this answers is not in the document]
Comment (fcanela) [reply src/services/email.ts:132]: The function should just do nothing when the transport is not configured.

File: src/utils/test-mocks.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { EventType, NotifyParams } from "../notify";
```
[the remark this answers is not in the document]
Comment (fcanela) [reply src/utils/test-mocks.ts:1]: Will apply in another PR. I have several imports to change for `import type`

File: src/notify.ts [hunk 4]
```diff
@@ -1,5 +1,7 @@
 import { LogNewAnswer, ProposalQuestionCreated } from "./services/reality";
```
[the remark this answers is not in the document]
Comment (fcanela) [reply src/notify.ts:1]: Will apply in another PR. I have several imports to change for `import type`

File: src/services/email.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (fcanela) [reply src/services/email.ts:132]: When a transport is not initialized, there is a warning in the logs with the missing env vars.

File: src/services/telegram.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+  if (missing.length > 0) {
+    emitter.emit(BotEventNames.TRANSPORT_CONFIGURATION_MISSING, {
+      name: TRANSPORT_NAME,
+      missing,
+    });
+    return;
+  }
+
+  client = new TelegramBot(env.TELEGRAM_TOKEN!, { polling: false });
```
[the remark this answers is not in the document]
Comment (fcanela) [reply src/services/telegram.ts:56]: The code snippet before this line is already ensuring it is not undefined. Sadly, TypeScript is not able to identify that and I require the ! symbol.

File: src/utils/tests-setup.ts [hunk 6]
```diff
@@ -1,10 +1,65 @@
 import chai from "chai";
 import chaiAsPromised from "chai-as-promised";
 import type { Address } from "viem";
+import { BotEventNames } from "../bot-events";
```
[the remark this answers is not in the document]
Comment (fcanela) [reply src/utils/tests-setup.ts:4]: What is the advantage of using unknown over any there to signify function arguments?

File: src/services/telegram.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (fcanela) [reply src/services/telegram.ts:56]: That code introduces extra branching to the already existing one. I prefer not adding duplicated tests for the same logic and placing the typescript `!` instead

File: src/services/stack.test.ts [hunk 7]
```diff
@@ -1,52 +1,44 @@
...
+  composeMessage,
+  configurableInitialize,
+  configurableNotify,
+} from "./slack";
 
 describe("Slack service", () => {
-  const fn = configurableInitialize;
   let emitter: EventEmitter;
 
   beforeEach(() => (emitter = new EventEmitter()));
```
[the remark this answers is not in the document]
Comment (fcanela) [reply src/services/stack.test.ts:19]: ?

Original code and suggested code is the same. Code doesn't seems to be used inside an expression.

File: src/utils/tests-setup.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (fcanela) [reply src/utils/tests-setup.ts]: Thanks, I will apply it in other PR.

[the remark this answers is not in the document]
Comment (fcanela) [reply src/utils/tests-setup.ts]: I have pending to do a PR to just add `import type` in general.

**Comment by fcanela** [comment]
@coderabbitai is there anything pending/blocking this PR from being merged on your side?

**Comment by fcanela** [comment]
@coderabbitai in that case, could you please mark the PR as "comments" or "approved" instead of changes requested or should I request another review for that to happen?

**Comment by fcanela** [comment]
@coderabbitai Dear AI tool,

You mentioned that you would mark the PR as approved but you didn't. While I completely trust the sincerity of machines, I am a little worried about the outcome. Would you mind indicating if there is any technical issue that impedes you from approving the PR? Is there any action on my side that I could preform to remove the unsettling red icon demanding changes from my side?

Regards,
Francisco

**Comment by jaybuidl** [comment]
@coderabbitai resolve

**Review by jaybuidl — APPROVED** [review]

**Pull request merged**